credentials/credentials_ext_test.go

```diff
@@ -0,0 +1,338 @@
+		t.Fatalf("EmptyCall() rpc failed: %v", err)
+	}
+}
+
+// FakeCredsNoAuthValidator is a test credential that does not implement AuthorityValidator.
+type FakeCredsNoAuthValidator struct {
+}
+
+// ClientHandshake performs the client-side handshake.
+func (c *FakeCredsNoAuthValidator) ClientHandshake(ctx context.Context, authority string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return rawConn, TestAuthInfo{}, nil
+}
+
+// TestAuthInfo implements the AuthInfo interface.
+type TestAuthInfo struct{}
+
+// AuthType returns the authentication type.
+func (TestAuthInfo) AuthType() string { return "test" }
+
+// Clone creates a copy of FakeCredsNoAuthValidator.
+func (c *FakeCredsNoAuthValidator) Clone() credentials.TransportCredentials {
+	return c
+}
+
+// Info provides protocol information.
+func (c *FakeCredsNoAuthValidator) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{}
+}
+
+// OverrideServerName overrides the server name used for verification.
+func (c *FakeCredsNoAuthValidator) OverrideServerName(serverName string) error {
+	return nil
+}
+
+// ServerHandshake performs the server-side handshake.
+// Returns a test AuthInfo object to satisfy the interface requirements.
+func (c *FakeCredsNoAuthValidator) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return rawConn, TestAuthInfo{}, nil
+}
+
+// FakeCredsWithAuthValidator is a test credential that does not implement AuthorityValidator.
+type FakeCredsWithAuthValidator struct {
+}
+
+// ClientHandshake performs the client-side handshake.
+func (c *FakeCredsWithAuthValidator) ClientHandshake(ctx context.Context, authority string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return rawConn, FakeAuthInfo{}, nil
+}
+
+// TestAuthInfo implements the AuthInfo interface.
+type FakeAuthInfo struct{}
+
+// AuthType returns the authentication type.
+func (FakeAuthInfo) AuthType() string { return "test" }
+
+// AuthType returns the authentication type.
+func (FakeAuthInfo) ValidateAuthority(authority string) error {
+	if authority == "auth.test.example.com" {
+		return nil
+	} else {
+		return fmt.Errorf("invalid authority")
+	}
+}
+
+// Clone creates a copy of FakeCredsWithAuthValidator.
+func (c *FakeCredsWithAuthValidator) Clone() credentials.TransportCredentials {
+	return c
+}
+
+// Info provides protocol information.
+func (c *FakeCredsWithAuthValidator) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{}
+}
+
+// OverrideServerName overrides the server name used for verification.
+func (c *FakeCredsWithAuthValidator) OverrideServerName(serverName string) error {
+	return nil
+}
+
+// ServerHandshake performs the server-side handshake.
+// Returns a test AuthInfo object to satisfy the interface requirements.
+func (c *FakeCredsWithAuthValidator) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
```

Comment:
Basically, we can have a single `testCreds` type that contains a boolean field that indicates whether or not it should return an `AuthInfo` that implements `AuthorityValidator` or not. We don't need two test credentials types here.